### DISCUSSION TITLE
How do I return an error from a custom action?

### DISCUSSION THREAD

**Discussion by lorg** [aminalaee/sqladmin#674] [Q&A] [answered]
Let's say I'm defining a custom action for my ModelView. How can I return an error that will be displayed nicely in the list view, and in the detail view?

**Comment by wnowicki** [comment]
I am actually creating custom action in my project at the moment, I've started round of some docs improvements https://github.com/aminalaee/sqladmin/pull/941. But I am facing the same issue, how to show user an error, especially useful in a single model actions.

**Comment by asko** [comment]
this can be done through combination of session data, context and changing [templates](https://github.com/aminalaee/sqladmin/blob/main/docs/working_with_templates.md):
for example, custom action in details can show response by e.g. 
1) adding logic to show the message similar to 
```jinja2
{% if action and action.result == 'error' %}
<div class="alert alert-danger" role="alert">{{ action.message }}</div>
{% endif %}
{% if action and action.result == 'success' %}
<div class="alert alert-success" role="alert">{{ action.message }}</div>
{% endif %}
```
2) subclassing Admin to override [`details`](https://github.com/aminalaee/sqladmin/blob/main/sqladmin/application.py#L470-L489) behavior (to extract action from e.g. request.session):
```python
    @login_required
    async def details(self, request: Request) -> Response:
        """Details route."""

        await self._details(request)
        model_view = self._find_model_view(request.path_params["identity"])
        model = await model_view.get_object_for_details(request)

        if not model:
            raise HTTPException(status_code=404)

        context = {
            "model_view": model_view,
            "model": model,
            "title": model_view.name,
        }
        session = request.session
        if session:
            context["action"] = session.pop("action", None)

        return await self.templates.TemplateResponse(
            request, model_view.details_template, context
        )
```
3) Implementing custom action at ModelView which would populate request.session.action with `result` and `message`:
```python
    @action(
        name="Show error",
        label="Show error",
        add_in_detail=True,
        add_in_list=True,
    )
    async def validate_key(self, request: Request):
        request.session["action"] = {
            "result": "error",
            "message": "Error message",
        }
        referer = request.headers.get("Referer")
        if referer:
            return RedirectResponse(referer)
        else:
            return RedirectResponse(
                request.url_for("admin:list", identity=self.identity)
            )
```

Alternatively one could extend `model_view` above to expose action results which can be further accessed through 
https://github.com/aminalaee/sqladmin/blob/4d230acd80d4016b23159e4249aa507973901b01/sqladmin/application.py#L481-L489 This could be more generic solution and fit both details and list actions (solution above only works for actions shown in details), but requires a bit more thought

**Comment by chezou** [comment]
I found that error template renders HTTPException message https://github.com/aminalaee/sqladmin/blob/4d230acd80d4016b23159e4249aa507973901b01/sqladmin/templates/sqladmin/error.html#L6-L7, and I used it:

```py
raise HTTPException(status_code=400, detail="Failed to the action")
```

**Comment by mmzeynalli** [comment, marked as answer]
Errors are already being automatically shown as of 0.25.0. Also, one can show custom `Flash` messages, if needed.
